kinematic_properties.py

```diff
@@ -242,24 +240,77 @@ def get_inertia_tensor(
        Masses of the particles.
      - position: unyt.unyt_array
        Positions of the particles.
+     - radius: unyt.unyt_quantity
+       Exclude particles outside this radius for the inertia tensor calculation
+     - reduced: bool
+       Whether to calculate the reduced inertia tensor
+     - max_iterations: int
+       The maximum number of iterations to repeat the inertia tensor calculation
 
     Returns the inertia tensor.
-    """
-    # 3x3 inertia tensor
-    Itensor = (
-        mass[:, None, None] / mass.sum() * position[:, None:, None] * position[:, None]
-    ).sum(axis=0)
-
-    # Symmetric, so only return lower triangle
-    Itensor = np.concatenate([np.diag(Itensor), Itensor[np.triu_indices(3, 1)]])
-
-    return Itensor
-
-
-def get_projected_inertia_tensor(
-    mass: unyt.unyt_array, position: unyt.unyt_array, axis: int
-) -> unyt.unyt_array:
-    """
+    '''
+    # Shouldn't calculate an inertia tensor for small numbers of particles
+    if position.shape[0] < 20:
+        return None
+
+    # Remove particles at centre if calculating reduced tensor
+    if reduced:
+        norm = np.linalg.norm(position, axis=1) ** 2
+        mask = np.logical_not(np.isclose(norm, 0))
+        position = position[mask]
+        mass = mass[mask]
+        norm = norm[mask]
+
+    # Set stopping criteria
+    tol = 0.0001
+    q = 1000
+
+    # Ensure we have consistent units
+    R = radius.to('kpc')
+    position = position.to('kpc')
+
+    # Start with a sphere
+    eig_val = [1, 1, 1]
+    eig_vec = np.array([
+        [1, 0, 0],
+        [0, 1, 0],
+        [0, 0, 1],
+    ])
+
+    for _ in range(max_iterations):
+        # Calculate shape
+        old_q = q
+        q = np.sqrt(eig_val[1] / eig_val[2])
+        s = np.sqrt(eig_val[0] / eig_val[2])
+        p = np.sqrt(eig_val[0] / eig_val[1])
+
+        # Break if converged
+        if abs((old_q - q) / q) < tol:
+            break
+
+        # Calculate ellipsoid, determine which particles are inside
+        axis = R * np.array([
+            1 * np.cbrt(s * p),
+            1 * np.cbrt(q / p),
+            1 / np.cbrt(q * s),
+        ])
+        p = np.dot(position, eig_vec) / axis
+        r = np.linalg.norm(p, axis=1)
+        weight = mass / np.sum(mass[r <= 1])
+        weight[r > 1] = 0
+
+        # Calculate inertia tensor
+        tensor = weight[:, None, None] * position[:, :, None] * position[:, None, :]
+        if reduced:
+            tensor /= norm[:, None, None]
+        tensor = tensor.sum(axis=0)
+        eig_val, eig_vec = np.linalg.eigh(tensor.value)
+
+    return np.concatenate([np.diag(tensor), tensor[np.triu_indices(3, 1)]])
+
+
+def get_projected_inertia_tensor(mass, position, axis, radius, reduced=False, max_iterations=20):
+    '''
     Get the inertia tensor of the given particle distribution, projected along the
```

Comment:
This can lead to confusion. As we discussed the other day, e.g. is it the inertia tensor of the projected particle distribution, or the projected inertia tensor of the particle distribution?  From our conversion, I think it is the former, but the comment makes me think of the latter.

Reply:
Updated in https://github.com/SWIFTSIM/SOAP/pull/82/commits/394ddd06eda98f6878b9906f5fb68193525b7c8a to (hopefully) clarify that it's the former